// @vscode-elements/elements/dist/vscode-button/vscode-button.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html, nothing } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { VscElement } from '../includes/VscElement.js';
import '../vscode-icon/index.js';
import styles from './vscode-button.styles.js';
import { ifDefined } from 'lit/directives/if-defined.js';
/**
 * @fires vsc-click Dispatched only when button is not in disabled state.
 *
 * @cssprop --vscode-button-background
 * @cssprop --vscode-button-foreground
 * @cssprop [--vscode-button-border=var(--vscode-button-background)]
 * @cssprop --vscode-button-hoverBackground
 * @cssprop --vscode-font-family
 * @cssprop --vscode-font-size
 * @cssprop --vscode-font-weight
 * @cssprop --vscode-button-secondaryForeground
 * @cssprop --vscode-button-secondaryBackground
 * @cssprop --vscode-button-secondaryHoverBackground
 * @cssprop --vscode-focusBorder
 */
let VscodeButton = class VscodeButton extends VscElement {
    get form() {
        return this._internals.form;
    }
    constructor() {
        super();
        this.autofocus = false;
        /** @internal */
        this.tabIndex = 0;
        /**
         * Button has a less prominent style.
         */
        this.secondary = false;
        /** @internal */
        this.role = 'button';
        this.disabled = false;
        /**
         * A [Codicon](https://microsoft.github.io/vscode-codicons/dist/codicon.html) before the label
         */
        this.icon = '';
        /**
         * Spin property for the icon
         */
        this.iconSpin = false;
        /**
         * A [Codicon](https://microsoft.github.io/vscode-codicons/dist/codicon.html) after the label
         */
        this.iconAfter = '';
        /**
         * Spin property for the after icon
         */
        this.iconAfterSpin = false;
        this.focused = false;
        this.name = undefined;
        this.type = 'button';
        this.value = '';
        this._prevTabindex = 0;
        this._handleFocus = () => {
            this.focused = true;
        };
        this._handleBlur = () => {
            this.focused = false;
        };
        this.addEventListener('keydown', this._handleKeyDown.bind(this));
        this.addEventListener('click', this._handleClick.bind(this));
        this._internals = this.attachInternals();
    }
    connectedCallback() {
        super.connectedCallback();
        if (this.autofocus) {
            if (this.tabIndex < 0) {
                this.tabIndex = 0;
            }
            this.updateComplete.then(() => {
                this.focus();
                this.requestUpdate();
            });
        }
        this.addEventListener('focus', this._handleFocus);
        this.addEventListener('blur', this._handleBlur);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.removeEventListener('focus', this._handleFocus);
        this.removeEventListener('blur', this._handleBlur);
    }
    update(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('value')) {
            this._internals.setFormValue(this.value);
        }
        if (changedProperties.has('disabled')) {
            if (this.disabled) {
                this._prevTabindex = this.tabIndex;
                this.tabIndex = -1;
            }
            else {
                this.tabIndex = this._prevTabindex;
            }
        }
    }
    _executeAction() {
        if (this.type === 'submit' && this._internals.form) {
            this._internals.form.requestSubmit();
        }
        if (this.type === 'reset' && this._internals.form) {
            this._internals.form.reset();
        }
    }
    _handleKeyDown(event) {
        if ((event.key === 'Enter' || event.key === ' ') &&
            !this.hasAttribute('disabled')) {
            /**
             * @deprecated
             * Please use the standard `click` event.
             */
            this.dispatchEvent(new CustomEvent('vsc-click', {
                detail: {
                    originalEvent: new MouseEvent('click'),
                },
            }));
            this._executeAction();
        }
    }
    _handleClick(event) {
        if (!this.hasAttribute('disabled')) {
            this.dispatchEvent(new CustomEvent('vsc-click', {
                detail: {
                    originalEvent: event,
                },
            }));
            this._executeAction();
        }
    }
    render() {
        const hasIcon = this.icon !== '';
        const hasIconAfter = this.iconAfter !== '';
        const wrapperClasses = {
            wrapper: true,
            'has-icon-before': hasIcon,
            'has-icon-after': hasIconAfter,
        };
        const iconElem = hasIcon
            ? html `<vscode-icon
          name="${this.icon}"
          ?spin="${this.iconSpin}"
          spin-duration="${ifDefined(this.iconSpinDuration)}"
          class="icon"
        ></vscode-icon>`
            : nothing;
        const iconAfterElem = hasIconAfter
            ? html `<vscode-icon
          name="${this.iconAfter}"
          ?spin="${this.iconAfterSpin}"
          spin-duration="${ifDefined(this.iconAfterSpinDuration)}"
          class="icon-after"
        ></vscode-icon>`
            : nothing;
        return html `
      <span class="${classMap(wrapperClasses)}">
        ${iconElem}
        <slot></slot>
        ${iconAfterElem}
      </span>
    `;
    }
};
VscodeButton.styles = styles;
/** @internal */
VscodeButton.formAssociated = true;
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeButton.prototype, "autofocus", void 0);
__decorate([
    property({ type: Number, reflect: true })
], VscodeButton.prototype, "tabIndex", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeButton.prototype, "secondary", void 0);
__decorate([
    property({ reflect: true })
], VscodeButton.prototype, "role", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeButton.prototype, "disabled", void 0);
__decorate([
    property()
], VscodeButton.prototype, "icon", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'icon-spin' })
], VscodeButton.prototype, "iconSpin", void 0);
__decorate([
    property({ type: Number, reflect: true, attribute: 'icon-spin-duration' })
], VscodeButton.prototype, "iconSpinDuration", void 0);
__decorate([
    property({ attribute: 'icon-after' })
], VscodeButton.prototype, "iconAfter", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'icon-after-spin' })
], VscodeButton.prototype, "iconAfterSpin", void 0);
__decorate([
    property({
        type: Number,
        reflect: true,
        attribute: 'icon-after-spin-duration',
    })
], VscodeButton.prototype, "iconAfterSpinDuration", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeButton.prototype, "focused", void 0);
__decorate([
    property({ type: String, reflect: true })
], VscodeButton.prototype, "name", void 0);
__decorate([
    property({ reflect: true })
], VscodeButton.prototype, "type", void 0);
__decorate([
    property()
], VscodeButton.prototype, "value", void 0);
VscodeButton = __decorate([
    customElement('vscode-button')
], VscodeButton);
export { VscodeButton };
